master/internal/scheduler/kubernetes_resource_provider.go

```diff
@@ -0,0 +1,252 @@
+
+	case AddTask:
+		k.receiveAddTask(ctx, msg)
+
+	case SetMaxSlots, SetWeight:
+
+	case SetTaskName:
+		k.receiveSetTaskName(ctx, msg)
+
+	case StartTask:
+		k.receiveStartTask(ctx, msg)
+
+	default:
+		ctx.Log().Errorf("Unexpected message %T", msg)
+		return actor.ErrUnexpectedMessage(ctx)
+	}
+
+	return nil
+}
+
+func (k *kubernetesResourceProvider) receiveAddTask(ctx *actor.Context, msg AddTask) {
+	actors.NotifyOnStop(ctx, msg.TaskHandler, taskStopped{Ref: msg.TaskHandler})
+
+	if task, ok := k.tasksByHandler[ctx.Sender()]; ok {
+		if ctx.ExpectingResponse() {
+			ctx.Respond(task)
+		}
+		return
+	}
+
+	if msg.Group == nil {
+		msg.Group = msg.TaskHandler
+	}
+	group := k.getOrCreateGroup(msg.Group, ctx)
+
+	var taskID TaskID
+	if msg.ID != nil {
+		taskID = *msg.ID
+	}
+
+	name := msg.Name
+	if len(name) == 0 {
+		name = "Unnamed-k8-Task"
+	}
+
+	task := newTask(&Task{
+		ID:                  taskID,
+		group:               group,
+		handler:             msg.TaskHandler,
+		name:                name,
+		slotsNeeded:         msg.SlotsNeeded,
+		canTerminate:        msg.CanTerminate,
+		agentLabel:          msg.Label,
+		fittingRequirements: msg.FittingRequirements,
+	})
+
+	k.tasksByID[task.ID] = task
+	k.tasksByHandler[task.handler] = task
+
+	if ctx.ExpectingResponse() {
+		ctx.Respond(task)
+	}
+
+	k.scheduleTask(ctx, task)
+}
+
+func (k *kubernetesResourceProvider) scheduleTask(ctx *actor.Context, task *Task) {
+	numPods := 1
+	slotsPerNode := task.SlotsNeeded()
+	if task.SlotsNeeded() > 1 {
+		if k.slotsPerNode == 0 {
+			ctx.Log().WithField("task ID", task.ID).Error(
+				"set slots_per_node > 0 to schedule tasks with slots")
+			return
+		}
+
+		if task.SlotsNeeded()%k.slotsPerNode != 0 {
+			ctx.Log().WithField("task ID", task.ID).Error(
+				"task number of slots is not schedulable on the configured slots_per_node")
+			return
+		}
+		numPods = task.SlotsNeeded() / k.slotsPerNode
+		slotsPerNode = k.slotsPerNode
+	}
+
+	for pod := 0; pod < numPods; pod++ {
+		k.assignPod(ctx, task, slotsPerNode)
+	}
+
+	ctx.Log().WithField("task ID", task.ID).Infof("task assigned by scheduler")
+	task.handler.System().Tell(task.handler, TaskAssigned{NumContainers: numPods})
+}
+
+func (k *kubernetesResourceProvider) assignPod(ctx *actor.Context, task *Task, slots int) {
+	if task.state != taskRunning {
+		task.mustTransition(taskRunning)
+	}
+	container := newContainer(task, k.agent, slots, len(task.containers))
+	k.agent.containers[container.id] = container
+	task.containers[container.id] = container
+	k.tasksByContainerID[container.id] = task
+	k.assigmentByTaskHandler[task.handler] = append(
+		k.assigmentByTaskHandler[task.handler],
+		podAssignment{
+			task:        task,
+			agent:       k.agent,
+			container:   container,
+			clusterID:   k.clusterID,
+			harnessPath: k.harnessPath,
+
+			taskContainerDefaults: k.taskContainerDefaults,
+		})
+}
+
+func (k *kubernetesResourceProvider) getOrCreateGroup(
+	handler *actor.Ref,
```

Comment:
non-blocking: Let's keep the context first everywhere it's passed as an argument.

Reply:
fixed